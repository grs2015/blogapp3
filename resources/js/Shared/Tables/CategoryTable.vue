<script setup lang="ts">

import { categoryData } from '@/Interfaces/PaginatedData'

interface Props {
    paginatedData: Array<categoryData>
}

const props = defineProps<Props>()

</script>

<template>
    CategoryTable
</template>
