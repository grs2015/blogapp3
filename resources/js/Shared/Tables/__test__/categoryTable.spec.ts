import { mount } from "@vue/test-utils";
import CategoryTable from "@/Shared/Tables/CategoryTable.vue"
import { i18nVue } from "laravel-vue-i18n";
import { Quasar } from "quasar";
import { Inertia } from '@inertiajs/inertia'

const mockGet = vi.spyOn(Inertia, 'get').mockImplementation(() => console.log('Sending Get request'))
const mockDelete = vi.spyOn(Inertia, 'delete').mockImplementation(() => console.log('Sending Delete request'))

let wrapper = null

const paginatedData = {
    current_page: 1,
    data: [
        { id: 1, title: 'first_title', color: 'purple', parent_id: 1, icon: 'description', slug: 'slug', content: 'content', meta_title: 'meta' },
        { id: 2, title: 'first_title', color: 'purple', parent_id: 1, icon: 'description', slug: 'slug', content: 'content', meta_title: 'meta' },
    ],
    first_page_url: "first_page_url",
    from: 1,
    last_page: 4,
    last_page_url: "last_page_url",
    links: [ { active: true, label: 'active', url: 'link' }, { active: false, label: 'inactive', url: 'link' } ],
    next_page_url: "next_page_url",
    path: "path",
    per_page: 5,
    prev_page_url: "prev_page_url",
    to: 4,
    total: 20
}
const sortingData = { column: 'title', descending: 'true' }

beforeEach(async() => {
    wrapper = await mount(CategoryTable,{ global: {
        plugins: [i18nVue, Quasar] },
        props: { paginatedData, sortingData },
        shallow: false })
    mockGet.mockReset()
    mockDelete.mockReset()
})

afterEach(async () => await wrapper.unmount());

describe('CategoryTable component', () => {
    it('checks the data from parent component is defined on current component', () => {
        expect(wrapper.element).toMatchSnapshot();

        expect(wrapper.props().paginatedData).toContain({
            current_page: 1,
            first_page_url: 'first_page_url',
            last_page_url: 'last_page_url',
            prev_page_url: 'prev_page_url',
        })
        expect(wrapper.props().paginatedData.data[0]).toContain({
            title: 'first_title',
            slug: 'slug',
            color: 'purple'
        })
    })

    it('checks the number of row in table (1 - Head, 2 - Body)', () => {
        expect(wrapper.findAll('tr')).toHaveLength(3)
    })

    it('checks the edit button events correcly hit the Inertia back-end', async () => {
        const editBtn = wrapper.find('[data-test="edit-button"]')

        await editBtn.trigger('click')
        expect(Inertia.get).toHaveBeenCalledTimes(1)
        expect(Inertia.get).toHaveBeenCalledWith("/admin/categories/slug/edit")
    })

    it('checks the delete button events correcly hit the Inertia back-end', async () => {
        const deleteBtn = wrapper.find('[data-test="delete-button"]')

        await deleteBtn.trigger('click')
        expect(Inertia.delete).toHaveBeenCalledTimes(1)
        expect(Inertia.delete).toHaveBeenCalledWith("/admin/categories/slug", expect.anything())
    })

    it('checks the refresh button events correcly hit the Inertia back-end', async () => {
        const refreshBtn = wrapper.find('[data-test="refresh-button"]')

        await refreshBtn.trigger('click')
        expect(Inertia.get).toHaveBeenCalledTimes(1)
        expect(Inertia.get).toHaveBeenCalledWith("/admin/categories")
    })

    it('checks the add button events correcly hit the Inertia back-end', async () => {
        const addBtn = wrapper.find('[data-test="add-button"]')

        await addBtn.trigger('click')
        expect(Inertia.get).toHaveBeenCalledTimes(1)
        expect(Inertia.get).toHaveBeenCalledWith("/admin/categories/create")
    })
})
