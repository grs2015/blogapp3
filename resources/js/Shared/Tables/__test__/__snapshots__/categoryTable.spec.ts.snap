// Vitest Snapshot v1

exports[`CategoryTable component > checks the data from parent component is defined on current component 1`] = `
<div
  class="q-mt-md"
>
  <div
    class="q-table__container q-table--horizontal-separator column no-wrap q-table__card q-table--flat q-table--bordered q-table--no-wrap"
  >
    <div
      class="q-table__top relative-position row items-center"
    >
      
      <div
        class="q-table__title text-primary"
      >
        Categories
      </div>
      <div
        class="q-space"
      />
      <label
        class="q-field row no-wrap items-start q-field--standard q-input q-field--labeled q-field--dense"
        for="f_879eaee4-cf5f-4e5c-9587-ed3b2fbbea3b"
      >
        <!---->
        <div
          class="q-field__inner relative-position col self-stretch"
        >
          <div
            class="q-field__control relative-position row no-wrap"
            tabindex="-1"
          >
            <div
              class="q-field__control-container col relative-position row no-wrap q-anchor--skip"
            >
              <input
                class="q-field__native q-placeholder"
                data-test="search-input"
                id="f_879eaee4-cf5f-4e5c-9587-ed3b2fbbea3b"
                tabindex="0"
                type="text"
              />
              <div
                class="q-field__label no-pointer-events absolute ellipsis text-primary"
              >
                <div>
                  Search
                </div>
              </div>
              <!---->
            </div>
            <div
              class="q-field__append q-field__marginal row no-wrap items-center"
            >
              <i
                aria-hidden="true"
                class="q-icon text-primary notranslate material-icons"
                role="presentation"
              >
                search
              </i>
            </div>
          </div>
          <!---->
        </div>
        <!---->
      </label>
      <button
        class="q-btn q-btn-item non-selectable no-outline q-btn--outline q-btn--rectangle text-primary q-btn--actionable q-focusable q-hoverable q-ml-md"
        data-test="refresh-button"
        tabindex="0"
        type="button"
      >
        <span
          class="q-focus-helper"
        />
        <span
          class="q-btn__content text-center col items-center q-anchor--skip justify-center row"
        >
          <i
            aria-hidden="true"
            class="q-icon notranslate material-icons"
            role="img"
          >
            restart_alt
          </i>
        </span>
        <transition-stub>
          <!---->
        </transition-stub>
      </button>
      <hr
        aria-orientation="vertical"
        class="q-separator q-separator--vertical q-separator--vertical-inset"
        style="margin-right: 8px; margin-left: 8px;"
      />
      <button
        class="q-btn q-btn-item non-selectable no-outline q-btn--unelevated q-btn--rectangle bg-green text-white q-btn--actionable q-focusable q-hoverable"
        data-test="add-button"
        tabindex="0"
        type="button"
      >
        <span
          class="q-focus-helper"
        />
        <span
          class="q-btn__content text-center col items-center q-anchor--skip justify-center row"
        >
          <i
            aria-hidden="true"
            class="q-icon on-left notranslate material-icons"
            role="presentation"
          >
            post_add
          </i>
          <div>
            ADD CATEGORY
          </div>
        </span>
      </button>
      
    </div>
    <div
      class="q-table__middle scroll"
    >
      <table
        class="q-table"
      >
        <thead>
          <tr
            class="q-tr"
          >
            
            <th
              class="text-center sortable text-primary"
            >
              Category Icon
              <i
                aria-hidden="true"
                class="q-icon notranslate material-icons q-table__sort-icon q-table__sort-icon--center"
                role="presentation"
              >
                arrow_upward
              </i>
            </th>
            <th
              class="text-left sortable text-primary"
            >
              Title
              <i
                aria-hidden="true"
                class="q-icon notranslate material-icons q-table__sort-icon q-table__sort-icon--left"
                role="presentation"
              >
                arrow_upward
              </i>
            </th>
            <th
              class="text-left sortable text-primary"
            >
              Content
              <i
                aria-hidden="true"
                class="q-icon notranslate material-icons q-table__sort-icon q-table__sort-icon--left"
                role="presentation"
              >
                arrow_upward
              </i>
            </th>
            <th
              class="text-center text-primary"
            >
              Actions
            </th>
            
          </tr>
        </thead>
        <tbody>
          <tr
            class=""
          >
            
            <td
              class="q-td q-table--col-auto-width text-center"
            >
              <i
                aria-hidden="true"
                class="q-icon text-purple notranslate material-icons"
                role="presentation"
                style="font-size: 32px;"
              >
                description
              </i>
            </td>
            
            <td
              class="text-left"
              style="width: 30%;"
            >
              first_title
            </td>
            <td
              class="text-left"
            >
              content
            </td>
            
            <td
              class="q-td q-table--col-auto-width text-center"
            >
              <div
                class="row flex-center q-gutter-x-sm no-wrap"
              >
                <button
                  class="q-btn q-btn-item non-selectable no-outline q-btn--outline q-btn--rectangle text-accent q-btn--actionable q-focusable q-hoverable"
                  data-test="edit-button"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="q-focus-helper"
                  />
                  <span
                    class="q-btn__content text-center col items-center q-anchor--skip justify-center row"
                  >
                    <i
                      aria-hidden="true"
                      class="q-icon notranslate material-icons"
                      role="img"
                    >
                      edit
                    </i>
                    <!---->
                  </span>
                </button>
                <button
                  class="q-btn q-btn-item non-selectable no-outline q-btn--outline q-btn--rectangle text-red q-btn--actionable q-focusable q-hoverable"
                  data-test="delete-button"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="q-focus-helper"
                  />
                  <span
                    class="q-btn__content text-center col items-center q-anchor--skip justify-center row"
                  >
                    <i
                      aria-hidden="true"
                      class="q-icon notranslate material-icons"
                      role="img"
                    >
                      delete
                    </i>
                    <!---->
                  </span>
                </button>
              </div>
            </td>
            
          </tr>
          <tr
            class=""
          >
            
            <td
              class="q-td q-table--col-auto-width text-center"
            >
              <i
                aria-hidden="true"
                class="q-icon text-purple notranslate material-icons"
                role="presentation"
                style="font-size: 32px;"
              >
                description
              </i>
            </td>
            
            <td
              class="text-left"
              style="width: 30%;"
            >
              first_title
            </td>
            <td
              class="text-left"
            >
              content
            </td>
            
            <td
              class="q-td q-table--col-auto-width text-center"
            >
              <div
                class="row flex-center q-gutter-x-sm no-wrap"
              >
                <button
                  class="q-btn q-btn-item non-selectable no-outline q-btn--outline q-btn--rectangle text-accent q-btn--actionable q-focusable q-hoverable"
                  data-test="edit-button"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="q-focus-helper"
                  />
                  <span
                    class="q-btn__content text-center col items-center q-anchor--skip justify-center row"
                  >
                    <i
                      aria-hidden="true"
                      class="q-icon notranslate material-icons"
                      role="img"
                    >
                      edit
                    </i>
                    <!---->
                  </span>
                </button>
                <button
                  class="q-btn q-btn-item non-selectable no-outline q-btn--outline q-btn--rectangle text-red q-btn--actionable q-focusable q-hoverable"
                  data-test="delete-button"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="q-focus-helper"
                  />
                  <span
                    class="q-btn__content text-center col items-center q-anchor--skip justify-center row"
                  >
                    <i
                      aria-hidden="true"
                      class="q-icon notranslate material-icons"
                      role="img"
                    >
                      delete
                    </i>
                    <!---->
                  </span>
                </button>
              </div>
            </td>
            
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="q-table__bottom row items-center justify-end"
    >
      <div
        class="q-table__separator col"
      />
      <div
        class="q-table__control"
      >
        <span
          class="q-table__bottom-item"
        >
          Records per page
        </span>
        <label
          class="q-field row no-wrap items-start q-field--borderless q-select q-field--auto-height q-select--without-input q-select--without-chips q-select--single q-field--float q-field--dense q-table__select inline q-table__bottom-item"
          for="f_124d6782-23b9-4eff-8c92-ba26569fac65"
        >
          <!---->
          <div
            class="q-field__inner relative-position col self-stretch"
          >
            <div
              class="q-field__control relative-position row no-wrap text-primary"
              tabindex="-1"
            >
              <div
                class="q-field__control-container col relative-position row no-wrap q-anchor--skip"
              >
                <div
                  class="q-field__native row items-center"
                >
                  <span>
                    5
                  </span>
                  <input
                    aria-autocomplete="none"
                    aria-controls="f_124d6782-23b9-4eff-8c92-ba26569fac65_lb"
                    aria-expanded="false"
                    aria-owns="f_124d6782-23b9-4eff-8c92-ba26569fac65_lb"
                    class="q-select__focus-target"
                    id="f_124d6782-23b9-4eff-8c92-ba26569fac65"
                    readonly=""
                    role="combobox"
                    tabindex="0"
                  />
                </div>
                <!---->
              </div>
              <div
                class="q-field__append q-field__marginal row no-wrap items-center q-anchor--skip"
              >
                <i
                  aria-hidden="true"
                  class="q-icon notranslate material-icons q-select__dropdown-icon"
                  role="presentation"
                >
                  arrow_drop_down
                </i>
              </div>
              <!---->
            </div>
            <!---->
          </div>
          <!---->
        </label>
      </div>
      <div
        class="q-table__control"
      >
        <button
          aria-disabled="true"
          class="q-btn q-btn-item non-selectable no-outline q-btn--flat q-btn--round text-grey-8 disabled q-btn--dense"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <span
            class="q-focus-helper"
          />
          <span
            class="q-btn__content text-center col items-center q-anchor--skip justify-center row"
          >
            <i
              aria-hidden="true"
              class="q-icon notranslate material-icons"
              role="img"
            >
              first_page
            </i>
          </span>
        </button>
        <button
          aria-disabled="true"
          class="q-btn q-btn-item non-selectable no-outline q-btn--flat q-btn--round text-grey-8 disabled q-btn--dense"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <span
            class="q-focus-helper"
          />
          <span
            class="q-btn__content text-center col items-center q-anchor--skip justify-center row"
          >
            <i
              aria-hidden="true"
              class="q-icon notranslate material-icons"
              role="img"
            >
              chevron_left
            </i>
          </span>
        </button>
        <button
          class="q-btn q-btn-item non-selectable no-outline q-btn--flat q-btn--round text-grey-8 q-btn--actionable q-focusable q-hoverable q-btn--dense"
          tabindex="0"
          type="button"
        >
          <span
            class="q-focus-helper"
          />
          <span
            class="q-btn__content text-center col items-center q-anchor--skip justify-center row"
          >
            <i
              aria-hidden="true"
              class="q-icon notranslate material-icons"
              role="img"
            >
              chevron_right
            </i>
          </span>
        </button>
        <button
          class="q-btn q-btn-item non-selectable no-outline q-btn--flat q-btn--round text-grey-8 q-btn--actionable q-focusable q-hoverable q-btn--dense"
          tabindex="0"
          type="button"
        >
          <span
            class="q-focus-helper"
          />
          <span
            class="q-btn__content text-center col items-center q-anchor--skip justify-center row"
          >
            <i
              aria-hidden="true"
              class="q-icon notranslate material-icons"
              role="img"
            >
              last_page
            </i>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
